File: scripts/sdk-validation/python/unipost_sdk_test.py
"""
unipost Python SDK Validation Test

Tests all supported API operations against the live UniPost API.

Setup:
  1. pip install -r requirements.txt
  2. Get an API key from https://app.unipost.dev → API Keys
  3. UNIPOST_API_KEY=up_live_xxx python unipost_sdk_test.py

Or with TEST_ACCOUNT_ID to test post creation:
  UNIPOST_API_KEY=up_live_xxx TEST_ACCOUNT_ID=<id> python unipost_sdk_test.py
"""

import os
import sys
import hmac
import hashlib
from datetime import datetime, timedelta, timezone
import requests as _requests

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../sdk/python")))

from unipost import UniPost, verify_webhook_signature

# ─── Config ───────────────────────────────────────────────────────────────────

API_KEY = os.environ.get("UNIPOST_API_KEY", "")
API_URL = os.environ.get("UNIPOST_API_URL", "https://api.unipost.dev")
TEST_ACCOUNT_ID = os.environ.get("TEST_ACCOUNT_ID", "")

# Track post IDs created during the test run for cleanup.
created_post_ids = []
created_webhook_ids = []

# ─── Test runner ──────────────────────────────────────────────────────────────

passed = 0
failed = 0
failures = []


def test(name, fn):
    global passed, failed
    print(f"  {name} ... ", end="", flush=True)
    try:
        result = fn()
        print("✅ PASS")
        passed += 1
        return result
    except Exception as e:
        print(f"❌ FAIL — {e}")
        failed += 1
        failures.append((name, str(e)))
        return None


def section(title):
    print(f"\n{'─' * 50}")
    print(f"  {title}")
    print("─" * 50)


# ─── Main ─────────────────────────────────────────────────────────────────────

def main():
    print("\n╔══════════════════════════════════════════════════╗")
    print("║     unipost (Python) — API Validation Test       ║")
    print("╚══════════════════════════════════════════════════╝\n")

    if not API_KEY:
        print("❌ Please set UNIPOST_API_KEY environment variable")
        print("   UNIPOST_API_KEY=up_live_xxx python unipost_sdk_test.py")
        sys.exit(1)

    client = UniPost(api_key=API_KEY)

    # ── 1. Accounts ────────────────────────────────────────────────────────────
    section("1. Accounts — list connected social accounts")

    accounts = test("accounts.list()", lambda: _test_list_accounts(client))

    if accounts:
        print(f"\n  Found {len(accounts)} connected accounts:")
        for a in accounts:
            name = a.account_name or a.id
            print(f"    • [{a.platform:<10}] {name}  (id: {a.id})")

        bluesky = next((a for a in accounts if a.platform == "bluesky"), None)
        if bluesky and not TEST_ACCOUNT_ID:
            print(f"\n  💡 Tip: Run with TEST_ACCOUNT_ID={bluesky.id} to test post creation")
        if not TEST_ACCOUNT_ID:
            candidate = bluesky or accounts[0]
            globals()["TEST_ACCOUNT_ID"] = candidate.id
            print(f"\n  Using TEST_ACCOUNT_ID={candidate.id} for safe draft/scheduled tests")

    # ── 2. Profiles — raw API smoke ────────────────────────────────────────────
    section("2. Profiles — list & filter accounts by profile")

    profiles = test("GET /v1/profiles", lambda: _test_profiles())
    if profiles:
        print(f"\n  Found {len(profiles)} profiles:")
        for p in profiles:
            print(f"    • {p['name']}  (id: {p['id']})")
        first = profiles[0]
        test(
            f"GET /v1/social-accounts?profile_id={first['id'][:8]}...",
            lambda: _test_profile_accounts(first["id"], first["name"]),
        )

    # ── 3. Webhooks — signature + CRUD ────────────────────────────────────────
    section("3. Webhooks — signature verification & subscription CRUD")

    test("verify_webhook_signature()", _test_verify_signature)

    webhook = test("webhooks.create()", lambda: client.webhooks.create(
        url="https://example.com/unipost-webhook-test",
        events=["post.published", "post.partial", "post.failed"],
    ))
    if webhook:
        created_webhook_ids.append(webhook.id)
        test("webhooks.list()", lambda: _test_list_webhooks(client, webhook.id))
        test(f'webhooks.get("{webhook.id[:8]}...")', lambda: _test_get_webhook(client, webhook.id))
        test("webhooks.update()", lambda: _test_update_webhook(client, webhook.id))
        test("webhooks.rotate()", lambda: _test_rotate_webhook(client, webhook.id))

    # ── 4. Posts — list & get ──────────────────────────────────────────────────
    section("4. Posts — list & get")

    posts = test("posts.list()", lambda: _test_list_posts(client))

    if posts and len(posts) > 0:
        first = posts[0]
        caption_preview = (first.caption or "")[:60]
        print(f'\n  First post: "{caption_preview}..."')
        print(f"  Status: {first.status}  |  Results: {len(first.results)}")

        test(f'posts.get("{first.id[:8]}...")', lambda: _test_get_post(client, first.id))
        test(f'posts.get_queue("{first.id[:8]}...")', lambda: _test_get_queue(client, first.id))
    else:
        print("\n  No posts yet — skipping posts.get() and posts.get_queue() tests")

    # ── 5. Posts — create ──────────────────────────────────────────────────────
    section("5. Posts — create (draft mode, no actual publishing)")

    if not TEST_ACCOUNT_ID:
        print("  ⏭  Skipped — set TEST_ACCOUNT_ID env var to run post creation tests")
    else:
        timestamp = datetime.now(timezone.utc).isoformat()
        caption = f"SDK validation test — {timestamp} [auto-generated]"

        draft = test("posts.create() — draft mode", lambda: _test_create_draft(client, caption))
        if draft:
            created_post_ids.append(draft.id)
            print(f"\n  Created draft post: {draft.id}")

        scheduled_at = (datetime.now(timezone.utc) + timedelta(minutes=10)).isoformat()
        scheduled = test(
            "posts.create() — scheduled mode",
            lambda: _test_create_scheduled(client, f"SDK scheduled — {timestamp}", scheduled_at),
        )
        if scheduled:
            created_post_ids.append(scheduled.id)
            print(f"  Created scheduled post: {scheduled.id}")
            test(
                f'posts.cancel("{scheduled.id[:8]}...")',
                lambda: _test_cancel(client, scheduled.id),
            )
            print("  Cancelled scheduled post ✓")

        publish_now = os.environ.get("TEST_PUBLISH_NOW") == "true"
        if publish_now:
            test(
                "posts.create() — publish NOW",
                lambda: _test_create_now(client, f"[SDK Test] Hello from unipost Python 🐍 {timestamp}"),
            )
        else:
            print("\n  ⏭  Real publish skipped (set TEST_PUBLISH_NOW=true to enable)")

    # ── 6. Analytics ───────────────────────────────────────────────────────────
    section("6. Analytics")

    now = datetime.now(timezone.utc)
    thirty_days_ago = now - timedelta(days=30)

    test(
        "analytics.rollup() — last 30 days",
        lambda: _test_analytics(client, thirty_days_ago.isoformat(), now.isoformat()),
    )

    # ── Cleanup ────────────────────────────────────────────────────────────────
    if created_webhook_ids or created_post_ids:
        section("7. Cleanup")
    for wid in created_webhook_ids:
        try:
            client.webhooks.delete(wid)
            print(f"  🧹 Deleted webhook {wid[:8]}...")
        except Exception as e:
            print(f"  ⚠  Failed to delete webhook {wid[:8]}... ({e})")
    if created_post_ids:
        for pid in created_post_ids:
            try:
                resp = _requests.delete(
                    f"{API_URL}/v1/social-posts/{pid}",
                    headers={"Authorization": f"Bearer {API_KEY}"},
                )
                resp.raise_for_status()
                print(f"  🗑  Deleted {pid[:8]}...")
            except Exception as e:
                print(f"  ⚠  Failed to delete {pid[:8]}... ({e})")

    # ── Summary ────────────────────────────────────────────────────────────────
    print(f"\n╔══════════════════════════════════════════════════╗")
    print(f"║  Results: {passed:2d} passed  {failed:2d} failed                    ║")
    print(f"╚══════════════════════════════════════════════════╝\n")

    if failed > 0:
        print("Failed tests:")
        for name, err in failures:
            print(f"  ❌ {name}: {err}")
        sys.exit(1)
    else:
        print("🎉 All tests passed! local unipost (Python) is working correctly.\n")


# ─── Test implementations ─────────────────────────────────────────────────────

def _test_list_accounts(client):
    res = client.accounts.list()
    data = res.get("data", [])
    if not isinstance(data, list):
        raise ValueError("Expected data array")
    return data


def _test_list_posts(client):
    res = client.posts.list(limit=5)
    data = res.get("data", [])
    if not isinstance(data, list):
        raise ValueError("Expected data array")
    return data


def _test_profiles():
    resp = _requests.get(
        f"{API_URL}/v1/profiles",
        headers={"Authorization": f"Bearer {API_KEY}"},
        timeout=30,
    )
    resp.raise_for_status()
    return resp.json().get("data", [])


def _test_profile_accounts(profile_id, profile_name):
    resp = _requests.get(
        f"{API_URL}/v1/social-accounts",
        params={"profile_id": profile_id},
        headers={"Authorization": f"Bearer {API_KEY}"},
        timeout=30,
    )
    resp.raise_for_status()
    data = resp.json().get("data", [])
    print(f'    → {len(data)} accounts in profile "{profile_name}"')
    return data


def _test_get_post(client, post_id):
    post = client.posts.get(post_id)
    if not post or not post.id:
        raise ValueError("Expected post with id")
    return post


def _test_get_queue(client, post_id):
    queue = client.posts.get_queue(post_id)
    if not queue or not queue.post or not queue.post.id:
        raise ValueError("Expected queue snapshot")
    return queue


def _test_create_draft(client, caption):
    post = client.posts.create(
        caption=caption,
        account_ids=[TEST_ACCOUNT_ID],
        status="draft",
    )
    if not post or not post.id:
        raise ValueError("Expected post with id")
    if post.status != "draft":
        raise ValueError(f"Expected status=draft, got {post.status}")
    return post


def _test_create_scheduled(client, caption, scheduled_at):
    post = client.posts.create(
        caption=caption,
        account_ids=[TEST_ACCOUNT_ID],
        scheduled_at=scheduled_at,
    )
    if not post or not post.id:
        raise ValueError("Expected post with id")
    if post.status != "scheduled":
        raise ValueError(f"Expected status=scheduled, got {post.status}")
    return post


def _test_cancel(client, post_id):
    post = client.posts.cancel(post_id)
    if not post:
        raise ValueError("Expected response")
    return post


def _test_create_now(client, caption):
    post = client.posts.create(
        caption=caption,
        account_ids=[TEST_ACCOUNT_ID],
    )
    if not post or not post.id:
        raise ValueError("Expected post with id")
    return post


def _test_analytics(client, from_date, to_date):
    rollup = client.analytics.rollup(
        from_date=from_date,
        to_date=to_date,
        granularity="day",
    )
    if not rollup:
        raise ValueError("Expected rollup data")
    return rollup


def _test_verify_signature():
    payload = b'{"event":"post.published","data":{"id":"post_test_123"}}'
    secret = "whsec_test_local"
    signature = "sha256=" + hmac.new(secret.encode("utf-8"), payload, hashlib.sha256).hexdigest()
    if not verify_webhook_signature(payload, signature, secret):
        raise ValueError("Expected signature to verify")
    return True


def _test_list_webhooks(client, webhook_id):
    payload = client.webhooks.list()
    data = payload.get("data", [])
    if not any(item.id == webhook_id for item in data):
        raise ValueError("Created webhook not found in list")
    return data


def _test_get_webhook(client, webhook_id):
    item = client.webhooks.get(webhook_id)
    if item.id != webhook_id:
        raise ValueError("Wrong webhook returned")
    if hasattr(item, "secret"):
        raise ValueError("Read response should not contain plaintext secret")
    return item


def _test_update_webhook(client, webhook_id):
    item = client.webhooks.update(webhook_id, active=False, events=["post.failed"])
    if item.active is not False:
        raise ValueError("Expected active=False")
    if item.events != ["post.failed"]:
        raise ValueError("Expected updated events")
    return item


def _test_rotate_webhook(client, webhook_id):
    item = client.webhooks.rotate(webhook_id)
    if not item.secret.startswith("whsec_"):
        raise ValueError("Expected rotated secret")
    return item


if __name__ == "__main__":
    main()

File: scripts/sdk-validation/js/unipost-sdk-test.mjs
/**
 * @unipost/sdk Validation Test
 *
 * Tests all supported API operations against the live UniPost API.
 *
 * Setup:
 *   1. npm install
 *   2. Get an API key from https://app.unipost.dev → API Keys
 *   3. UNIPOST_API_KEY=up_live_xxx node unipost-sdk-test.mjs
 *
 * Or set directly below:
 */

import crypto from 'node:crypto';
import { UniPost, verifyWebhookSignature } from '../../../sdk/javascript/dist/index.mjs';

// ─── Config ───────────────────────────────────────────────────────────────────

const API_KEY = process.env.UNIPOST_API_KEY || 'YOUR_API_KEY_HERE';
const API_URL = process.env.UNIPOST_API_URL || 'https://api.unipost.dev';

// After running listAccounts(), paste one account_id here for post tests.
// Leave empty to skip post creation tests (safe for first run).
let TEST_ACCOUNT_ID = process.env.TEST_ACCOUNT_ID || '';

// Track post IDs created during the test run for cleanup.
const createdPostIds = [];
const createdWebhookIds = [];

// ─── Test runner ──────────────────────────────────────────────────────────────

let passed = 0;
let failed = 0;
const results = [];

async function test(name, fn) {
  process.stdout.write(`  ${name} ... `);
  try {
    const result = await fn();
    console.log('✅ PASS');
    passed++;
    results.push({ name, status: 'pass', data: result });
    return result;
  } catch (err) {
    console.log(`❌ FAIL — ${err.message}`);
    failed++;
    results.push({ name, status: 'fail', error: err.message });
    return null;
  }
}

function section(title) {
  console.log(`\n${'─'.repeat(50)}`);
  console.log(`  ${title}`);
  console.log('─'.repeat(50));
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log('\n╔══════════════════════════════════════════════════╗');
  console.log('║     @unipost/sdk — API Validation Test           ║');
  console.log('╚══════════════════════════════════════════════════╝\n');

  if (API_KEY === 'YOUR_API_KEY_HERE') {
    console.error('❌ Please set UNIPOST_API_KEY environment variable');
    console.error('   UNIPOST_API_KEY=up_live_xxx node unipost-sdk-test.mjs');
    process.exit(1);
  }

  const client = new UniPost({ apiKey: API_KEY });

  // ── 1. Accounts ─────────────────────────────────────────────────────────────
  section('1. Accounts — list connected social accounts');

  const accounts = await test('listAccounts()', async () => {
    const res = await client.accounts.list();
    if (!res || !Array.isArray(res.data)) throw new Error('Expected data array');
    if (!res.meta || typeof res.meta.total !== 'number') throw new Error('Expected meta.total');
    if (typeof res.meta.limit !== 'number') throw new Error('Expected meta.limit');
    if (res.data.length === 0) throw new Error('No accounts found — connect at least one');
    return res.data;
  });

  if (accounts) {
    console.log(`\n  Found ${accounts.length} connected accounts:`);
    accounts.forEach(a => {
      console.log(`    • [${a.platform.padEnd(10)}] ${a.account_name || a.id}  (id: ${a.id})`);
    });

    // Print first Bluesky account id as a good safe test target
    const bluesky = accounts.find(a => a.platform === 'bluesky');
    if (bluesky && !TEST_ACCOUNT_ID) {
      console.log(`\n  💡 Tip: Run with TEST_ACCOUNT_ID=${bluesky.id} to test post creation`);
    }
  }

  if (!TEST_ACCOUNT_ID && accounts?.length) {
    const safestAccount = accounts.find(a => a.platform === 'bluesky') || accounts[0];
    TEST_ACCOUNT_ID = safestAccount.id;
    console.log(`\n  Using TEST_ACCOUNT_ID=${TEST_ACCOUNT_ID} for safe draft/scheduled tests`);
  }

  // ── 2. Profiles — list (raw API, pending SDK v0.2.0) ──────────────────────
  section('2. Profiles — list & filter accounts by profile');

  const profilesRes = await test('GET /v1/profiles', async () => {
    const res = await fetch(`${API_URL}/v1/profiles`, {
      headers: { Authorization: `Bearer ${API_KEY}` },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const json = await res.json();
    if (!Array.isArray(json.data)) throw new Error('Expected data array');
    return json.data;
  });

  if (profilesRes && profilesRes.length > 0) {
    console.log(`\n  Found ${profilesRes.length} profiles:`);
    profilesRes.forEach(p => console.log(`    • ${p.name}  (id: ${p.id})`));

    // Test profile_id filter on accounts
    const firstProfile = profilesRes[0];
    await test(`GET /v1/social-accounts?profile_id=${firstProfile.id.slice(0, 8)}...`, async () => {
      const res = await fetch(`${API_URL}/v1/social-accounts?profile_id=${firstProfile.id}`, {
        headers: { Authorization: `Bearer ${API_KEY}` },
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const json = await res.json();
      if (!Array.isArray(json.data)) throw new Error('Expected data array');
      console.log(`    → ${json.data.length} accounts in profile "${firstProfile.name}"`);
      return json.data;
    });
  }

  // ── 3. Webhooks — signature + CRUD ─────────────────────────────────────────
  section('3. Webhooks — signature verification & subscription CRUD');

  await test('verifyWebhookSignature()', async () => {
    const payload = JSON.stringify({
      event: 'post.published',
      timestamp: new Date().toISOString(),
      data: { id: 'post_test_123' },
    });
    const secret = 'whsec_test_local';
    const signature = `sha256=${crypto.createHmac('sha256', secret).update(payload).digest('hex')}`;
    const valid = await verifyWebhookSignature({ payload, signature, secret });
    if (!valid) throw new Error('Expected signature to verify');
    return true;
  });

  const webhook = await test('webhooks.create()', async () => {
    const res = await client.webhooks.create({
      url: 'https://example.com/unipost-webhook-test',
      events: ['post.published', 'post.partial', 'post.failed'],
    });
    if (!res?.id) throw new Error('Expected webhook id');
    if (!res.secret || !res.secret.startsWith('whsec_')) throw new Error('Expected generated signing secret');
    return res;
  });

  if (webhook?.id) {
    createdWebhookIds.push(webhook.id);

    await test('webhooks.list()', async () => {
      const res = await client.webhooks.list();
      if (!Array.isArray(res?.data)) throw new Error('Expected data array');
      if (!res.data.find((item) => item.id === webhook.id)) throw new Error('Created webhook not found in list');
      if (!res.meta || typeof res.meta.total !== 'number') throw new Error('Expected meta.total');
      if (typeof res.meta.limit !== 'number') throw new Error('Expected meta.limit');
      return res.data;
    });

    await test(`webhooks.get("${webhook.id.slice(0, 8)}...")`, async () => {
      const res = await client.webhooks.get(webhook.id);
      if (res.id !== webhook.id) throw new Error('Wrong webhook returned');
      if ('secret' in res) throw new Error('Read response should not contain plaintext secret');
      return res;
    });

    await test('webhooks.update()', async () => {
      const res = await client.webhooks.update(webhook.id, {
        active: false,
        events: ['post.failed'],
      });
      if (res.active !== false) throw new Error('Expected active=false');
      if (!Array.isArray(res.events) || res.events.length !== 1 || res.events[0] !== 'post.failed') {
        throw new Error('Expected updated events');
      }
      return res;
    });

    await test('webhooks.rotate()', async () => {
      const res = await client.webhooks.rotate(webhook.id);
      if (!res.secret || !res.secret.startsWith('whsec_')) throw new Error('Expected rotated secret');
      return res;
    });
  }

  // ── 4. Posts — list ─────────────────────────────────────────────────────────
  section('4. Posts — list & get');

  const posts = await test('posts.list()', async () => {
    const res = await client.posts.list({ limit: 5 });
    if (!res || !Array.isArray(res.data)) throw new Error('Expected data array');
    if (res.meta && res.meta.next_cursor !== undefined && res.nextCursor !== res.meta.next_cursor) {
      throw new Error('Expected nextCursor to mirror meta.next_cursor');
    }
    return res.data;
  });

  if (posts && posts.length > 0) {
    const firstPost = posts[0];
    console.log(`\n  First post: "${(firstPost.caption || '').slice(0, 60)}..."`);
    console.log(`  Status: ${firstPost.status}  |  Platform results: ${firstPost.results?.length ?? 0}`);

    await test(`posts.get("${firstPost.id.slice(0, 8)}...")`, async () => {
      const res = await client.posts.get(firstPost.id);
      if (!res?.id) throw new Error('Expected post object with id');
      return res;
    });

    await test(`posts.getQueue("${firstPost.id.slice(0, 8)}...")`, async () => {
      const res = await client.posts.getQueue(firstPost.id);
      if (!res?.post?.id) throw new Error('Expected queue snapshot with post');
      if (!Array.isArray(res.jobs)) throw new Error('Expected jobs array');
      return res;
    });
  } else {
    console.log('\n  No posts yet — skipping posts.get() and posts.getQueue() tests');
  }

  // ── 5. Posts — create (only if TEST_ACCOUNT_ID is set) ──────────────────────
  section('5. Posts — create (draft mode, no actual publishing)');

  if (!TEST_ACCOUNT_ID) {
    console.log('  ⏭  Skipped — set TEST_ACCOUNT_ID env var to run post creation tests');
    console.log('     Example: TEST_ACCOUNT_ID=<account_id> node unipost-sdk-test.mjs');
  } else {
    const timestamp = new Date().toISOString();
    const caption = `SDK validation test — ${timestamp} [auto-generated, will be deleted]`;

    // Test: create a draft (safe, not actually published)
    let draftPostId = null;
    const draft = await test('posts.create() — draft mode', async () => {
      const res = await client.posts.create({
        caption,
        accountIds: [TEST_ACCOUNT_ID],
        status: 'draft',
      });
      if (!res?.id) throw new Error('Expected post object with id');
      if (res.status !== 'draft') throw new Error(`Expected status=draft, got ${res.status}`);
      return res;
    });

    if (draft) {
      draftPostId = draft.id;
      createdPostIds.push(draftPostId);
      console.log(`\n  Created draft post: ${draftPostId}`);
    }

    // Test: create a scheduled post (10 min in future)
    const scheduledAt = new Date(Date.now() + 10 * 60 * 1000).toISOString();
    let scheduledPostId = null;
    const scheduled = await test('posts.create() — scheduled mode (10 min future)', async () => {
      const res = await client.posts.create({
        caption: `SDK scheduled test — ${timestamp}`,
        accountIds: [TEST_ACCOUNT_ID],
        scheduledAt,
      });
      if (!res?.id) throw new Error('Expected post object with id');
      if (res.status !== 'scheduled') throw new Error(`Expected status=scheduled, got ${res.status}`);
      return res;
    });

    if (scheduled) {
      scheduledPostId = scheduled.id;
      createdPostIds.push(scheduledPostId);
      console.log(`  Created scheduled post: ${scheduledPostId}`);
      console.log(`  Scheduled for: ${scheduledAt}`);
    }

    // Test: cancel the scheduled post
    if (scheduledPostId) {
      await test(`posts.cancel("${scheduledPostId.slice(0, 8)}...")`, async () => {
        const res = await client.posts.cancel(scheduledPostId);
        if (!res) throw new Error('Expected response data');
        return res;
      });
      console.log('  Cancelled scheduled post ✓');
    }

    // Test: create NOW post (real publish — only if explicitly confirmed)
    const publishNow = process.env.TEST_PUBLISH_NOW === 'true';
    if (publishNow) {
      await test('posts.create() — publish NOW (real post)', async () => {
        const res = await client.posts.create({
          caption: `[SDK Test] Hello from local @unipost/sdk 🚀 ${timestamp}`,
          accountIds: [TEST_ACCOUNT_ID],
        });
        if (!res?.id) throw new Error('Expected post object with id');
        return res;
      });
    } else {
      console.log('\n  ⏭  Real publish skipped (set TEST_PUBLISH_NOW=true to enable)');
    }
  }

  // ── 6. Analytics ────────────────────────────────────────────────────────────
  section('6. Analytics');

  const now = new Date();
  const thirtyDaysAgo = new Date(now.getTime() - 30 * 24 * 60 * 60 * 1000);

  await test('analytics.rollup() — last 30 days', async () => {
    const res = await client.analytics.rollup({
      from: thirtyDaysAgo.toISOString(),
      to: now.toISOString(),
      granularity: 'day',
    });
    if (!res) throw new Error('Expected rollup data');
    return res;
  });

  // ── Cleanup ─────────────────────────────────────────────────────────────────
  if (createdWebhookIds.length > 0 || createdPostIds.length > 0) {
    section('7. Cleanup');
  }

  if (createdWebhookIds.length > 0) {
    for (const id of createdWebhookIds) {
      try {
        await client.webhooks.delete(id);
        console.log(`  🧹 Deleted webhook ${id.slice(0, 8)}...`);
      } catch {
        console.log(`  ⚠  Failed to delete webhook ${id.slice(0, 8)}... (non-fatal)`);
      }
    }
  }

  if (createdPostIds.length > 0) {
    for (const id of createdPostIds) {
      try {
        await fetch(`${API_URL}/v1/social-posts/${id}`, {
          method: 'DELETE',
          headers: { Authorization: `Bearer ${API_KEY}` },
        });
        console.log(`  🗑  Deleted ${id.slice(0, 8)}...`);
      } catch {
        console.log(`  ⚠  Failed to delete ${id.slice(0, 8)}... (non-fatal)`);
      }
    }
  }

  // ── Summary ──────────────────────────────────────────────────────────────────
  console.log('\n╔══════════════════════════════════════════════════╗');
  console.log(`║  Results: ${String(passed).padStart(2)} passed  ${String(failed).padStart(2)} failed                    ║`);
  console.log('╚══════════════════════════════════════════════════╝\n');

  if (failed > 0) {
    console.log('Failed tests:');
    results
      .filter(r => r.status === 'fail')
      .forEach(r => console.log(`  ❌ ${r.name}: ${r.error}`));
    process.exit(1);
  } else {
    console.log('🎉 All tests passed! local @unipost/sdk is working correctly.\n');
  }
}

main().catch(err => {
  console.error('\n💥 Unexpected error:', err);
  process.exit(1);
});
